struct.js

```diff
@@ -0,0 +1,316 @@
+        );
+        if (self.strict && self.isArgument && !field.required) {
+            assert.ok(false, 'every field must be marked required on ' + self.name +
+                ' including ' + field.name + ' in strict mode');
+        }
+        if (self.isArgument && field.optional) {
+            assert.ok(false, 'no field of an argument struct may be marked optional including ' + field.name +
+                ' of ' + self.name);
+        }
+        field.required = field.required || self.isArgument;
+
+        // Field names must be valid JavaScript. If the Thrift name is not
+        // valid in JavaScript, it can be overridden with the js.name
+        // annotation.
+        field.name = field.annotations && field.annotations['js.name'] || field.name;
+        self.fieldsById[field.id] = field;
+        self.fieldsByName[field.name] = field;
+        self.fieldNames[index] = field.name;
+        self.fields.push(field);
+    }
+};
+
+StructSpec.prototype.link = function link(spec) {
+    var self = this;
+    self.Constructor = self.createConstructor(self.name, self.fieldNames);
+    self.Constructor.rw = self.rw;
+    spec[self.name] = self.Constructor;
+    for (var index = 0; index < self.fields.length; index++) {
+        self.fields[index].link(spec);
+    }
+};
+
+// The following methods have alternate implementations for Exception and Union.
+
+StructSpec.prototype.createConstructor = function createConstructor(name, fieldNames) {
+    var source;
+    source = '(function $' + name + '(options) {\n';
+    for (var index = 0; index < fieldNames.length; index++) {
+        var fieldName = fieldNames[index];
+        source += '    this.' + fieldName + ' = null;\n';
+        source += 'if (options && \'' + fieldName + '\' in options) ' +
+            '{ this.' + fieldName + ' = options.' + fieldName + '; }';
+    }
+    source += '})\n';
+    // eval is an operator that captures the lexical scope of the calling
+    // function and deoptimizes the lexical scope.
+    // By using eval in an expression context, it loses this second-class
+    // capability and becomes a first-class function.
+    // (0, eval) is one way to use eval in an expression context.
+    return (0, eval)(source);
+};
+
+StructSpec.prototype.initialize = function initialize() {
+    var self = this;
+    return new self.Constructor();
+};
+
+StructSpec.prototype.set = function set(struct, key, value) {
+    struct[key] = value;
+};
+
+StructSpec.prototype.finalize = function finalize(struct) {
+    return struct;
+};
+
+function StructRW(spec) {
+    assert(spec, 'spec required');
+    var self = this;
+    self.spec = spec;
+}
+
+StructRW.prototype.byteLength = function byteLength(struct) {
+    var self = this;
+    var spec = self.spec;
+    var length = 1; // stop:1
+    var result;
+    for (var index = 0; index < spec.fields.length; index++) {
+        var field = spec.fields[index];
+        var value = struct[field.name];
```

Comment:
I’m punting default value. It’s on the WIP workboard at #15.

If we want to accept `null` as a valid struct body (per @jwolski’s request for arguments) this may not be necessary and is definitely not sufficient.